src/data_processing/data_processing.py

```diff
@@ -74,101 +66,80 @@ def deduplicate(df):
 
-    df["AwardAbstract"] = np.where(
-        df["AwardAbstract"].str.strip().str.lower().isin(abstract_nulls),
-        "",
-        df["AwardAbstract"],
+    Args:
+        df(pd.DataFrame): Dataset containing AwardTitle and AwardAbstract str columns.
+
+    Returns:
+        pd.DataFrame: Returns dataframe with cleaned and combined AllText str column.
+    
+    """
+    ref = config[text_type]
+    df = df.loc[~df[col].str.strip().str.lower().isin(ref["drop"])]
+    df[col] = df[col].fillna("")
+    
+    df[col] = np.where(
+        df[col].str.strip().str.lower().isin(ref["nulls"]), "", df[col]
     )
 
-    # Remove grants that start with 'pivotal nurse support contract'
-    df["AwardTitle"] = df["AwardTitle"].fillna("")
-    df = df.loc[
-        ~df["AwardTitle"].str.lower().str.startswith("pivotal nurse support contract")
-    ]
-
-    # Remove grants that start with 'nihr in-practice fellowship'
-    df = df.loc[
-        ~df["AwardTitle"].str.lower().str.startswith("nihr in-practice fellowship")
-    ]
-
-    # Remove grants that start with 'nurture: national unified renal'
-    df = df.loc[
-        ~df["AwardTitle"].str.lower().str.startswith("nurture: national unified renal")
-    ]
-
-    # Remove grants that start with ''qi project: assist-ckd'
-    df = df.loc[~df["AwardTitle"].str.lower().str.startswith("qi project: assist-ckd")]
-
-    # Remove common funder specific boiler plate prefixes from abstracts.
-    for term in ("background", "background,", "background:", "objectives"):
-        # ToDo: Replace with regex.
-        df["AwardAbstract"] = np.where(
-            df["AwardAbstract"].str[: len(term)].str.lower() == term,
-            df["AwardAbstract"].str[len(term) :],
-            df["AwardAbstract"],
-        )
-
-    df["AwardAbstract"] = df["AwardAbstract"].str.strip()
+    for prefix in ref["prefixes"]:
+        df = df.loc[~df[col].str.lower().str.startswith(prefix)]
+
+    for pat in ref["boiler_plate"]:
+        df[col] = df[col].str.replace(f"^{pat}", "", n=1, case=False, regex=True)
+
+    removal_chars = string.punctuation + string.whitespace
+    df[col] = df[col].str.lstrip(removal_chars)
+    df[col] = df[col].str.replace(r'\s+', ' ', regex=True)
+    df[col] = df[col].str.strip()
+
+    return df
+
+
+def process_texts(df, min_char_len=110):
+    """Clean and combine title and abstract texts.
+
+    Args:
+        df(pd.DataFrame): Dataset containing AwardTitle and AwardAbstract str columns.
+
+    Returns:
+        pd.DataFrame: Returns dataframe with cleaned and combined AllText str column.
+    
+    """
+    df = clean_text_column(df.copy(), "AwardTitle", "title")
+    df = clean_text_column(df.copy(), "AwardAbstract", "abstract")
+
     df["AllText"] = df["AwardTitle"].fillna("") + " " + df["AwardAbstract"].fillna("")
-    # df.drop(columns=['AwardTitle', 'AwardAbstract'], inplace=True)
-    df = df.loc[df["AllText"].str.len() >= 110].copy()
+    df["AllText"] = df["AllText"].str.strip()
```

Comment:
`str.strip()` removes any leading or trailing whitespaces, this however retains whitespaces anywhere else. 

I found that `" ".join(text.split())` is very good at removing any whitespaces that are anywhere else. Also: by default, .split() separates at whitespace, including spaces, tabs, and newlines.